src/libopenrave/kinbodystatesaver.cpp

```diff
@@ -18,6 +18,121 @@
+                if( !pGrabbingLink ) {
+                    throw OPENRAVE_EXCEPTION_FORMAT(_("env=%s, could not find grabbing link for grabbed body '%s'"),
+                                                    GetEnv()->GetNameId()%pGrabbedBody->GetName(),
+                                                    ORE_Failed);
+                }
+                else if( !GetLink(pGrabbingLink->GetName()) ) {
+                    throw OPENRAVE_EXCEPTION_FORMAT(_("env=%s, could not find grabbing link '%s' for grabbed body '%s' since %s does not have the link (body num links is %d)"),
+                                                    GetEnv()->GetNameId()%pGrabbingLink->GetName()%pGrabbedBody->GetName()%GetName()%GetLinks().size(),
+                                                    ORE_Failed);
+                }
+                if( GetEnv() == savedBody.GetEnv() ) {
+                    // attach and add
+                    _AttachBody(pGrabbedBody);
+                    BOOST_ASSERT(pGrabbedBody->GetEnvironmentBodyIndex() > 0);
+                    _grabbedBodiesByEnvironmentIndex[pGrabbedBody->GetEnvironmentBodyIndex()] = pGrabbed;
+
+                    // grabbed bodies could have been removed from env and self collision checker.
+                    CollisionCheckerBasePtr collisionchecker = GetSelfCollisionChecker();
+                    if (!!collisionchecker) {
+                        collisionchecker->InitKinBody(pGrabbedBody);
+                    }
+                }
+                else {
+                    // The body that the state was saved from is from a different environment from savedBody. This case can
+                    // happen when cloning an environment (see Environment::_Clone). Since cloning is supposed to
+                    // preserve environmentBodyIndex of bodies, it is ok do the following.
+
+                    KinBodyPtr pNewGrabbedBody = GetEnv()->GetBodyFromEnvironmentBodyIndex(pGrabbedBody->GetEnvironmentBodyIndex());
+                    if( !!pNewGrabbedBody ) {
+                        if( pGrabbedBody->GetKinematicsGeometryHash() != pNewGrabbedBody->GetKinematicsGeometryHash() ) {
+                            RAVELOG_WARN_FORMAT("env=%s, new grabbed body '%s' kinematics-geometry hash is different from original grabbed body '%s' from env=%s", GetEnv()->GetNameId()%pNewGrabbedBody->GetName()%pGrabbedBody->GetName()%savedBody.GetEnv()->GetNameId());
+                        }
+                        else {
+                            // body is supposed to already be set to some "proper" configuration as the newly
+                            // initialized Grabbed objects will save the current state of pbody for later computation of
+                            // _listNonCollidingLinksWhenGrabbed (in case it is not yet computed).
+                            KinBody::LinkPtr pNewGrabbingLink = GetLinks().at(KinBody::LinkPtr(pGrabbingLink)->GetIndex());
```

Comment:
What's the cast for?

Reply:
I guess this is same question as https://github.com/rdiankov/openrave/pull/1429#discussion_r1767769482?

Reply:
No I mean why is there an explicit `KinBody::LinkPtr()` around pGrabbingLink? Isn't it already a linkptr? Won't this just create another shared_ptr just to delete it again?

Reply:
Oh, right!

The code was originally there, but indeed, it's unnecessary.

I also find one more place that creates unnecessary shared_ptr.

Fixed via https://github.com/rdiankov/openrave/pull/1429/commits/d53538ed7bf238aab574e9c69fb5dad856c4cc99, thanks!